x/go/calc/calc.go

```diff
@@ -0,0 +1,209 @@
+				tokens = append(tokens, currToken)
+				currToken = ""
+			}
+			continue
+		}
+		//	Check if token is a negative number
+		if c == '-' && currToken == "" && (i == 0 || tokens[len(tokens)-1] == "(" || tokens[len(tokens)-1] == "+" || tokens[len(tokens)-1] == "-" || tokens[len(tokens)-1] == "*" || tokens[len(tokens)-1] == "/" || tokens[len(tokens)-1] == "^") {
+			currToken += string(c)
+		} else if c == '+' || c == '-' || c == '*' || c == '/' || c == '(' || c == ')' || c == '^' {
+			if currToken != "" {
+				tokens = append(tokens, currToken)
+				currToken = ""
+			}
+			tokens = append(tokens, string(c))
+		} else {
+			currToken += string(c)
+		}
+	}
+	if currToken != "" {
+		tokens = append(tokens, currToken)
+	}
+	return tokens, nil
+}
+
+// Given a string representing a valid mathematical expression, sets exp to the
+// parsed expression
+func (e *Expression) Build(s string) error {
+	tokens, err := findTokens(s)
+	if err != nil {
+		return err
+	}
+	//	Uses the shunting yard algorithm to create an ast
+	precedence := map[string]int{
+		"+": 1,
+		"-": 1,
+		"*": 2,
+		"/": 2,
+		"(": 0,
+		"^": 3,
+	}
+	tokensDict := map[string]token.Token{
+		"+": token.ADD,
+		"-": token.SUB,
+		"*": token.MUL,
+		"/": token.QUO,
+		"^": token.XOR,
+	}
+	output := stack.Stack[interface{}]{}
+	operators := stack.Stack[string]{}
+
+	for _, t := range tokens {
+		switch t {
+		case "+", "-", "*", "/":
+			for operators.Len() > 0 && precedence[*operators.Peek()] >= precedence[t] {
+				op, err := operators.Pop()
+				if err != nil {
+					return err
+				}
+				Y, err1 := output.Pop()
+				X, err2 := output.Pop()
+				if err1 != nil || err2 != nil {
+					return errors.New("Invalid expression")
+				}
+				output.Push(&ast.BinaryExpr{X: X.(ast.Expr), Op: tokensDict[op], Y: Y.(ast.Expr)})
+			}
+			operators.Push(t)
+		case "^":
+			for operators.Len() > 0 && precedence[*operators.Peek()] > precedence[t] {
+				op, err := operators.Pop()
+				if err != nil {
+					return err
+				}
+				Y, err1 := output.Pop()
+				X, err2 := output.Pop()
+				if err1 != nil || err2 != nil {
+					return errors.New("Invalid expression")
+				}
+				output.Push(&ast.BinaryExpr{X: X.(ast.Expr), Op: tokensDict[op], Y: Y.(ast.Expr)})
+			}
+			operators.Push(t)
+		case "(":
+			operators.Push(t)
+		case ")":
+			for operators.Len() > 0 && *operators.Peek() != "(" {
+				op, err := operators.Pop()
+				if err != nil {
+					return err
+				}
+				Y, err1 := output.Pop()
+				X, err2 := output.Pop()
+				if err1 != nil || err2 != nil {
+					return errors.New("Invalid expression")
+				}
+				output.Push(&ast.BinaryExpr{X: X.(ast.Expr), Op: tokensDict[op], Y: Y.(ast.Expr)})
+			}
+			_, err := operators.Pop()
+			if err != nil {
+				return err
+			}
+		default:
+			_, err := strconv.ParseFloat(t, 64)
+			if err != nil {
+				output.Push(&ast.Ident{Name: t})
+			} else {
+				output.Push(&ast.BasicLit{Kind: token.FLOAT, Value: t})
+			}
+
+		}
+	}
+	for operators.Len() > 0 {
```

Comment:
It seems like this is a close math with the code from L119 to L134.  Would it make sense to extract into a separate method?